File: blog/clojure-python-c-three-programming-languages-one-algorithm/naive_compute_primes.clj
;; Clojure implementation of a naive prime numbers calculation algorithm.
;;
;; On my system (a Thinkpad X250 with an Intel i5-5300U CPU)
;; it takes ~12500ms to compute 10000 prime numbers.

(defn next-prime
  "Given a vector of all the prime numbers up to N in ascending order, return the next
  prime number."
  [primes]
  (loop [n (+ 2  (last primes))]
    (if (some zero? (map #(mod n %) primes))
      (recur (+ 2 n))
      n)))

(defn extend-primes
  "Given a vector of all the prime numbers up to N in ascending order extend it
  adding the next prime to the end of the vector."
  [primes]
  (conj primes (next-prime primes)))

(defn primes
  "Get the first N primes."
  [N] ((apply comp (repeat (- N 2) extend-primes)) [2 3]))

File: blog/clojure-python-c-three-programming-languages-one-algorithm/naive_compute_primes.py
"""Python implementation of a naive prime numbers calculation algorithm.

On my system (a Thinkpad X250 with an Intel i5-5300U CPU)
it takes ~3000ms to compute 10000 prime numbers.
"""


def primes(num_primes):
    """Compute the first `num_primes` prime numbers."""
    primes = [2, 3]

    for i in range(2, num_primes):
        n = primes[-1] + 2
        while True:
            if all(n % p for p in primes):
                primes.append(n)
                break
            n += 2

    return primes
